boa/src/class.rs

```diff
@@ -108,157 +107,81 @@ impl<T: Class> ClassConstructor for T {
 }
 
 impl<'context> ClassBuilder<'context> {
     pub(crate) fn new<T>(context: &'context mut Context) -> Self
     where
         T: ClassConstructor,
     {
-        let global = context.global_object();
-
-        let prototype = {
-            let object_prototype = global.get_field("Object").get_field(PROTOTYPE);
-
-            let object = Object::create(object_prototype);
-            GcObject::new(object)
-        };
-        // Create the native function
-        let function = Function::BuiltIn(
-            BuiltInFunction(T::raw_constructor),
-            FunctionFlags::CONSTRUCTABLE,
-        );
-
-        // Get reference to Function.prototype
-        // Create the function object and point its instance prototype to Function.prototype
-        let mut constructor =
-            Object::function(function, global.get_field("Function").get_field(PROTOTYPE));
-
-        let length = Property::data_descriptor(
-            T::LENGTH.into(),
-            Attribute::READONLY | Attribute::NON_ENUMERABLE | Attribute::PERMANENT,
-        );
-        constructor.insert_property("length", length);
-
-        let name = Property::data_descriptor(
-            T::NAME.into(),
-            Attribute::READONLY | Attribute::NON_ENUMERABLE | Attribute::PERMANENT,
-        );
-        constructor.insert_property("name", name);
-
-        let constructor = GcObject::new(constructor);
-
-        prototype
-            .borrow_mut()
-            .insert_field("constructor", constructor.clone().into());
-
-        constructor
-            .borrow_mut()
-            .insert_field(PROTOTYPE, prototype.clone().into());
-
-        Self {
-            context,
-            object: constructor,
-            prototype,
-        }
+        let mut builder = ConstructorBuilder::new(context, T::raw_constructor);
+        builder.name(T::NAME);
+        builder.length(T::LENGTH);
+        Self { builder }
     }
 
-    pub(crate) fn build(self) -> GcObject {
-        self.object
+    pub(crate) fn build(mut self) -> GcObject {
+        self.builder.build()
     }
 
     /// Add a method to the class.
     ///
     /// It is added to `prototype`.
-    pub fn method<N>(&mut self, name: N, length: usize, function: NativeFunction)
+    pub fn method<N>(&mut self, name: N, length: usize, function: NativeFunction) -> &mut Self
     where
         N: Into<String>,
     {
-        let name = name.into();
-        let mut function = Object::function(
-            Function::BuiltIn(function.into(), FunctionFlags::CALLABLE),
-            self.context
-                .global_object()
-                .get_field("Function")
-                .get_field("prototype"),
-        );
-
-        function.insert_field("length", Value::from(length));
-        function.insert_field("name", Value::from(name.as_str()));
-
-        self.prototype
-            .borrow_mut()
-            .insert_field(name, Value::from(function));
+        // TODO: \/
```

Comment:
Can you detail the `TODO` or was it from when you were implementing it?

Reply:
> Can you detail the `TODO` or was it from when you were implementing it?

It was something I wanted to implement, I'll do it now :)